c2rust/src/main.rs

```diff
@@ -1,56 +1,154 @@
+            })
+    }
+
+    /// Get all known ([`Self::known`]) and actual, found ([`Self::find_all`]) subcommands,
+    /// putting the known ones first so that the found ones overwrite them and take precedence.
+    pub fn all() -> anyhow::Result<impl Iterator<Item = Self>> {
+        Ok(Self::known().chain(Self::find_all()?))
+    }
+
+    pub fn invoke<I, S>(&self, args: I) -> anyhow::Result<()>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let path = self.path.as_ref().ok_or_else(|| {
+            anyhow!(
+                "known subcommand not found (probably not built): {}",
+                self.name
+            )
+        })?;
+        let status = Command::new(&path).args(args).status()?;
+        process::exit(status.code().unwrap_or(1));
+    }
 }
 
-fn invoke_subcommand<I, S>(subcommand: &str, args: I)
-where
-    I: IntoIterator<Item = S>,
-    S: AsRef<OsStr>,
-{
-    // Assumes the subcommand executable is in the same directory as this driver
-    // program.
-    let cmd_path = std::env::current_exe().expect("Cannot get current executable path");
-    let mut cmd_path = cmd_path.as_path().canonicalize().unwrap();
-    cmd_path.pop(); // remove current executable
-    cmd_path.push(format!("c2rust-{}", subcommand));
-    assert!(cmd_path.exists(), "{:?} is missing", cmd_path);
-    exit(
-        Command::new(cmd_path.into_os_string())
-            .args(args)
-            .status()
-            .expect("SubCommand failed to start")
-            .code()
-            .unwrap_or(-1),
-    );
+fn main() -> anyhow::Result<()> {
+    let sub_commands = SubCommand::all()?.collect::<Vec<_>>();
+    let sub_commands = sub_commands
+        .iter()
+        .map(|cmd| (cmd.name.as_ref(), cmd))
+        .collect::<HashMap<_, _>>();
+
+    // If the subcommand matches, don't use `clap` at all.
+    //
+    // I can't seem to get `clap` to pass through all arguments as is,
+    // like the ones with hyphens like `--metadata`,
+    // even though I've set [`Arg::allow_hyphen_values`].
+    // This is faster anyways.
+    // I also tried a single "subcommand" argument with [`Arg::possible_values`],
+    // but that had the same problem passing through all arguments as well.
+    //
+    // Furthermore, doing it this way correctly forwards `--help` through to the subcommand
+    // instead of `clap` intercepting it and displaying the top-level `--help`.
+    let mut args = env::args_os();
+    args.next(); // skip exe name
+    let sub_command = args.next();
+    let sub_command = sub_command
+        .as_ref()
+        .and_then(|arg| arg.to_str())
+        .and_then(|name| sub_commands.get(name));
+
+    if let Some(sub_command) = sub_command {
+        return sub_command.invoke(args);
+    }
+
+    // If we didn't get a subcommand, then use `clap` for parsing and error/help messages.
+    let matches = App::new("C2Rust")
+        .version(&*render_testament!(TESTAMENT))
```

Comment:
Not for this change, but maybe we should add this version thing to the other subcommands. I don't see it for c2rust-transpile at least.

Reply:
Yeah, that might be a good idea.